app/controllers/incomeSources/add/AddIncomeSourceStartDateCheckController.scala

```diff
@@ -177,39 +175,63 @@ class AddIncomeSourceStartDateCheckController @Inject()(authenticate: Authentica
       case (Some(form.responseYes), SelfEmployment) =>
-        Redirect(successUrl)
-          .addingToSession(
-            SessionKeys.addBusinessAccountingPeriodStartDate -> incomeSourceStartDate,
-            SessionKeys.addBusinessAccountingPeriodEndDate -> dateService.getAccountingPeriodEndDate(incomeSourceStartDate)
-          )
+        setSessionData(Seq(
+          (addBusinessAccountingPeriodStartDate, incomeSourceStartDate),
+          (addBusinessAccountingPeriodEndDate, dateService.getAccountingPeriodEndDate(incomeSourceStartDate))
+        )
+          , Redirect(successUrl)
+        ) map {
+          case Left(_) => showInternalServerError(isAgent)
+          case Right(result) => result
+        }
       case (Some(form.responseYes), _) =>
-        Redirect(successUrl)
+        Future.successful(Redirect(successUrl))
       case _ =>
         Logger("application").error(s"[AddIncomeSourceStartDateCheckController][handleValidForm] - Unexpected response, isAgent = $isAgent")
-        showInternalServerError(isAgent)
+        Future.successful(showInternalServerError(isAgent))
     }
   }
 
-  private def getAndValidateStartDate(incomeSourceType: IncomeSourceType)
-                                     (implicit user: MtdItUser[_]): Either[Throwable, String] = {
+  private def removeSessionData(keys: Seq[String], successResult: Result, errorResult: Result)(implicit mtdItUser: MtdItUser[_]): Future[Result] = {
+    sessionService.remove(keys, successResult) map {
+      case Left(_) => errorResult
+      case Right(result) => result
+    }
+  }
+
+  private def setSessionData(keyValuePairs: Seq[(String, String)], result: Result)(implicit mtdItUser: MtdItUser[_]): Future[Either[Throwable, Result]] = {
+    keyValuePairs match {
+      case x :: xs =>
+        sessionService.set(x._1, x._2, result) flatMap {
+          case Right(addedResult) => setSessionData(xs, addedResult)
+          case Left(ex) => Future.successful(Left(ex))
+        }
+      case _ => Future.successful(Right(result))
+    }
+  }
 
-    val maybeIncomeSourceStartDate = user.session.get(incomeSourceType.startDateSessionKey)
+  private def getStartDate(incomeSourceType: IncomeSourceType)(implicit user: MtdItUser[_]): Future[Either[Throwable, Option[String]]] = {
+    sessionService.get(incomeSourceType.startDateSessionKey)
+  }
 
-    maybeIncomeSourceStartDate match {
-      case None => Left(new Error(s"Session value not found for Key: ${incomeSourceType.startDateSessionKey}"))
-      case Some(date) if Try(date.toLocalDate).toOption.isDefined => Right(date)
-      case Some(invalidDate) => Left(new Error(s"Could not parse: $invalidDate as LocalDate for Key: ${incomeSourceType.startDateSessionKey}"))
+  private def getAndValidateStartDate(incomeSourceType: IncomeSourceType)
+                                     (implicit user: MtdItUser[_]): Future[Either[Throwable, String]] = {
+
+    getStartDate(incomeSourceType) map {
```

Comment:
you would really simplify this code if you mode attempt to convert to LocalDate into getStartDate method which is one liner at the moment